File: CppProject/gfg/inversionCount.cpp
#include <iostream>
#include <vector>

using namespace std;

long long mergeAndCount(vector<int>& arr, int l, int m, int r) {
    int size_A = m - l + 1;
    int size_B = r - m;

    vector<int> A(size_A);
    vector<int> B(size_B);

    for (int i = 0; i < size_A; i++) {
        A[i] = arr[l + i];
    }

    for (int i = 0; i < size_B; i++) {
        B[i] = arr[m + 1 + i];
    }

    int i = 0, j = 0, k = l;
    long long inversions = 0;

    while (i < size_A && j < size_B) {
        if (A[i] <= B[j]) {
            arr[k] = A[i];
            i++;
        } else {
            arr[k] = B[j];
            inversions += (size_A - i); // Count the inversions
            j++;
        }
        k++;
    }

    while (i < size_A) {
        arr[k] = A[i];
        i++;
        k++;
    }

    while (j < size_B) {
        arr[k] = B[j];
        j++;
        k++;
    }

    return inversions;
}

long long mergeSortAndCount(vector<int>& arr, int l, int r) {
    long long inversions = 0;
    if (l < r) {
        int m = l + (r - l) / 2;

        inversions += mergeSortAndCount(arr, l, m);
        inversions += mergeSortAndCount(arr, m + 1, r);
        inversions += mergeAndCount(arr, l, m, r);
    }
    return inversions;
}

int main() {
    vector<int> arr = {2, 4, 1, 3, 5};
    int n = arr.size();

    long long inversionCount = mergeSortAndCount(arr, 0, n - 1);
    cout << "Number of inversions are: " << inversionCount << endl;

    return 0;
}

File: PythonProject/gfg/inversionCount.py
def merge_and_count(arr, l, m, r):
    size_A = m - l + 1
    size_B = r - m
    
    A = arr[l:l + size_A]
    B = arr[m + 1:m + 1 + size_B]
    
    i = 0
    j = 0
    k = l
    inversions = 0
    
    while i < size_A and j < size_B:
        if A[i] <= B[j]:
            arr[k] = A[i]
            i += 1
        else:
            arr[k] = B[j]
            inversions += size_A - i
            j += 1
        k += 1
    
    while i < size_A:
        arr[k] = A[i]
        i += 1
        k += 1
    
    while j < size_B:
        arr[k] = B[j]
        j += 1
        k += 1
    
    return inversions

def merge_sort_and_count(arr, l, r):
    if l >= r:
        return 0
    
    m = (l + r) // 2
    inversion_count = 0
    
    inversion_count += merge_sort_and_count(arr, l, m)
    inversion_count += merge_sort_and_count(arr, m + 1, r)
    inversion_count += merge_and_count(arr, l, m, r)
    
    return inversion_count

def inversion_count(arr):
    n = len(arr)
    return merge_sort_and_count(arr, 0, n - 1)

arr = [1, 20, 6, 4, 5]
print("Number of inversions are", inversion_count(arr))